### PR TITLE
#143 Action/Code Generation. Cron job generation

### PR DESCRIPTION
**Pull request by roma-glushko**
### Description
This PR unlocks Magento cronjob generation inside of PHPStorm Plugin. Here is a brief demo of the functionality:

Feature List:
- generates Cronjob Classes and register them in crontab.xml
- adds a new cronjob action
- adds a new cronjob dialog
- adds the cron group indexer to be able to select group for a new cronjob
- suggests cronjob name based on the cronjob classname and module name
- supports the predefined list of commonly used cron schedules (like every minute or at midnight)
- supports fixed and configurable cron schedules
- all code generation templates are configurable and adjustable
- adds cron schedule expression and config path validations

### Fixed Issues
1. magento/magento2-phpstorm-plugin#143: Action/Code Generation. Cron job generation

### Contribution checklist 
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages

### CODE REVIEW

File: resources/fileTemplates/code/Magento Crontab Cronjob Registration.xml.html [hunk 1]
```diff
@@ -0,0 +1,54 @@
+<!--
+/*
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+-->
```
Comment (VitaliyBoyko) [inline resources/fileTemplates/code/Magento Crontab Cronjob Registration.xml.html:6]: Hi @roma-glushko 
We are not storing copyrights under templates. Could you please remove them. Thank you!

File: src/com/magento/idea/magento2plugin/actions/generation/generator/CronjobClassGenerator.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+    /**
+     *
+     * @param actionName
+     *
+     * @return void
+     */
+    public PsiFile generate(String actionName) {
+        String errorTitle = "Error";
+//        String cronjobFqn = this.getCronjobFqn();
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/generator/CronjobClassGenerator.java:54]: Probably you forget some to remove this :)

File: src/com/magento/idea/magento2plugin/actions/generation/generator/CrontabXmlGenerator.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+import com.magento.idea.magento2plugin.magento.files.CrontabXmlTemplate;
+import com.magento.idea.magento2plugin.util.xml.XmlPsiTreeUtil;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Properties;
+
+/**
+ *
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/generator/CrontabXmlGenerator.java:31]: Redundant doc block.

File: src/com/magento/idea/magento2plugin/actions/generation/generator/CrontabXmlGenerator.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        }
+
+        for (XmlTag cronjobTag: cronjobTags) {
+            if (!cronjobTag.getName().equals(CrontabXmlTemplate.CRON_JOB_TAG)) {
+                continue;
+            }
+
+            XmlAttribute[] cronjobAttributes = PsiTreeUtil.getChildrenOfType(cronjobTag, XmlAttribute.class);
+
+            // todo: handle null pointer
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/generator/CrontabXmlGenerator.java:166]: No todos, let's do it right away :)

File: src/com/magento/idea/magento2plugin/actions/generation/generator/CrontabXmlGenerator.java (hunk 3 above)
Comment (roma-glushko) [reply src/com/magento/idea/magento2plugin/actions/generation/generator/CrontabXmlGenerator.java]: Removed ✅

File: src/com/magento/idea/magento2plugin/actions/generation/generator/CronjobClassGenerator.java (hunk 2 above)
Comment (roma-glushko) [reply src/com/magento/idea/magento2plugin/actions/generation/generator/CronjobClassGenerator.java]: Refactored ✅

File: src/com/magento/idea/magento2plugin/actions/generation/generator/CrontabXmlGenerator.java (hunk 4 above)
Comment (roma-glushko) [reply src/com/magento/idea/magento2plugin/actions/generation/generator/CrontabXmlGenerator.java:176]: haha, yeah, added exception handling 🙌

File: resources/fileTemplates/code/Magento Crontab Cronjob Registration.xml.html (hunk 1 above)
Comment (roma-glushko) [reply resources/fileTemplates/code/Magento Crontab Cronjob Registration.xml.html:6]: @VitaliyBoyko it seems like a template explanation, not exactly the template. We used to add copyrights to all of such files. Please let me know if this changes the situation here 🙌

Comment (VitaliyBoyko) [reply resources/fileTemplates/code/Magento Crontab Cronjob Registration.xml.html:6]: Sorry, sure

**Comment by VitaliyBoyko** [comment]
Good to be merged ✅
image

**Pull request merged**